src/signalrclient/hub_connection_impl.cpp

```diff
@@ -510,6 +518,114 @@ namespace signalr
         m_disconnected = disconnected;
     }
 
+    void hub_connection_impl::reset_send_ping()
+    {
+        auto timeMs = (std::chrono::steady_clock::now() + m_signalr_client_config.get_keepalive_interval()).time_since_epoch();
+        m_nextActivationSendPing.store(std::chrono::duration_cast<std::chrono::milliseconds>(timeMs).count());
+    }
+
+    void hub_connection_impl::reset_server_timeout()
+    {
+        auto timeMs = (std::chrono::steady_clock::now() + m_signalr_client_config.get_server_timeout()).time_since_epoch();
+        m_nextActivationServerTimeout.store(std::chrono::duration_cast<std::chrono::milliseconds>(timeMs).count());
+    }
+
+    void hub_connection_impl::start_keepalive(std::weak_ptr<hub_connection_impl> weak_connection)
```

Comment:
This looks like it will always be called from an active connection so we don't need to accept a `weak_ptr` and check it in this method.